airflow/models.py

```diff
@@ -3536,11 +3536,14 @@ def set(
         session.expunge_all()
 
         # remove any duplicate XComs
-        session.query(cls).filter(
+        x = session.query(cls).filter(
             cls.key == key,
             cls.execution_date == execution_date,
             cls.task_id == task_id,
-            cls.dag_id == dag_id).delete()
+            cls.dag_id == dag_id).first()
+
+        if x:
+            session.delete(x)
```

Comment:
Why can't we update? Why do we need to do a delete followed by an add? It seems like a better option would be something along the lines below that you might find documented on http://docs.sqlalchemy.org/en/latest/core/dml.html
```
cls.update().\
            where(
              cls.key == key,		              
              cls.execution_date == execution_date,		          
              cls.task_id == task_id,
              cls.dag_id == dag_id).\
              values(f_1='v1')
```